allow installing rest to existing webservice

pweil- opened the pull request:
Allow passing a web service to `InstallREST` and subsequently `Install` to allow conditionally registering REST implementations to a pre-existing web service/api prefix.  If a nil web service is detected a new web service will be created.

@liggitt @smarterclayton @pmorie

bgrant0607 commented:
cc @kubernetes/goog-csi

lavalamp commented:
What is the purpose of this?

@brendandburns seems like maybe some overlap with your recent/planned work.

smarterclayton commented:
This was to allow us to inject legacy support rest objects into an existing
we service

On Sep 11, 2015, at 6:01 PM, Daniel Smith <notifications@github.com> wrote:

What is the purpose of this?

@brendandburns <https://github.com/brendandburns> seems like maybe some
overlap with your recent/planned work.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/13525#issuecomment-139671011>
.

lavalamp commented on pkg/apiserver/api_installer.go:
```diff
@@ -82,10 +79,10 @@ func (a *APIInstaller) Install() (ws *restful.WebService, errors []error) {
 }
 
-func (a *APIInstaller) newWebService() *restful.WebService {
+func (a *APIInstaller) NewWebService() *restful.WebService {
```
Please add godoc.

pweil- replied:
updated

lavalamp commented on pkg/apiserver/apiserver.go:
```diff
@@ -113,8 +113,11 @@ const (
-// in a slash. A restful WebService is created for the group and version.
-func (g *APIGroupVersion) InstallREST(container *restful.Container) error {
+// in a slash.
+// If you are installing the API group to an existing webservice then it should be passed in.  If
```
What do you think about adding a separate Update or Reinstall method, instead of making this method do two things? Perhaps it could even find the correct web service inside the container? Is there a reason to make the caller find the web service?

pweil- replied:
Hmm, I initially started out with a two method approach but switched just due to duplication.  Perhaps I can consolidate it nicely.  Will take another pass at that.

lavalamp replied:
I'm OK if one calls the other under the hood, but I think keeping the public interface clear is super important.

pweil- commented:
@lavalamp round 2 - added `UpdateREST` that will find the web service based on the installer's prefix and reverted the signature change for `InstallREST`.  Also cleaned up the test.

Commit:
allow installing rest to existing web service

lavalamp commented:
LGTM
